sw/device/tests/sim_dv/spi_tpm_smoketest.c

```diff
@@ -0,0 +1,217 @@
+      kDifSpiDeviceIrqUploadPayloadNotEmpty,
+      kDifSpiDeviceIrqUploadPayloadOverflow,
+      kDifSpiDeviceIrqReadbufWatermark,
+      kDifSpiDeviceIrqReadbufFlip,
+      kDifSpiDeviceIrqTpmHeaderNotEmpty};
+
+  for (uint32_t i = 0; i <= ARRAYSIZE(spi_device_irqs); ++i) {
+    CHECK_DIF_OK(dif_spi_device_irq_set_enabled(spi_device, spi_device_irqs[i],
+                                                kDifToggleEnabled));
+  }
+}
+
+void ottf_external_isr(void) {
+  plic_isr_ctx_t plic_ctx = {.rv_plic = &plic,
+                             .hart_id = kTopEarlgreyPlicTargetIbex0};
+
+  spi_device_isr_ctx_t spi_device_ctx = {
+      .spi_device = &spi_device.dev,
+      .plic_spi_device_start_irq_id =
+          kTopEarlgreyPlicIrqIdSpiDeviceGenericRxFull,
+      .expected_irq = kDifSpiDeviceIrqTpmHeaderNotEmpty,
+      .is_only_irq = true};
+
+  top_earlgrey_plic_peripheral_t peripheral;
+  dif_spi_device_irq_t spi_device_irq;
+  isr_testutils_spi_device_isr(plic_ctx, spi_device_ctx, &peripheral,
+                               &spi_device_irq);
+
+  switch (spi_device_irq) {
+    case kDifSpiDeviceIrqTpmHeaderNotEmpty:
+      header_interrupt_received = true;
+      // Disable interrupt until work is handled.
+      CHECK_DIF_OK(dif_spi_device_irq_set_enabled(
+          &spi_device.dev, kDifSpiDeviceIrqTpmHeaderNotEmpty,
+          kDifToggleDisabled));
+      break;
+    default:
+      LOG_ERROR("Unexpected interrupt: %d", spi_device_irq);
+      break;
+  }
+}
+
+static void ack_spi_tpm_header_irq(dif_spi_device_handle_t *spi_device) {
+  // Clear interrupt state and re-enable interrupt.
+  CHECK_DIF_OK(dif_spi_device_irq_acknowledge(
+      &spi_device->dev, kDifSpiDeviceIrqTpmHeaderNotEmpty));
+  CHECK_DIF_OK(dif_spi_device_irq_set_enabled(
+      &spi_device->dev, kDifSpiDeviceIrqTpmHeaderNotEmpty, kDifToggleEnabled));
+}
+
+bool test_main(void) {
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+
+  CHECK_DIF_OK(dif_spi_device_init_handle(
+      mmio_region_from_addr(TOP_EARLGREY_SPI_DEVICE_BASE_ADDR), &spi_device));
+
+  CHECK_DIF_OK(dif_rv_plic_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR), &plic));
+
+  // Set IoA7 for tpm csb.
+  // Longer term this needs to migrate to a top specific, platform specific
+  // setting.
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInSpiDeviceTpmCsb,
+      kTopEarlgreyPinmuxInselIoa7));
+
+  CHECK_DIF_OK(
+      dif_spi_device_tpm_configure(&spi_device, kDifToggleEnabled, tpm_config));
+
+  // enable interrupts
+  en_plic_irqs(&plic);
+  en_spi_device_irqs(&spi_device.dev);
+
+  // Sync message with testbench to begin.
+  LOG_INFO("Begin TPM Test");
+
+  for (uint32_t i = 0; i < kIterations; i++) {
+    LOG_INFO("Iteration %d", i);
+
+    // Wait for write interrupt.
+    while (!header_interrupt_received) {
+      wait_for_interrupt();
+    }
```

Comment:
Oops, I don't think this is correct use of the `wfi` instruction. As you have probably noticed, this construction has a race condition between `wfi` and the interrupt you want to wait on. The key to avoiding races is in this blurb from the privileged spec:

```
The operation of WFI
must be unaffected by the global interrupt bits in mstatus (MIE and SIE) and the delegation
register mideleg (i.e., the hart must resume if a locally enabled interrupt becomes pending, even if
it has been delegated to a less-privileged mode), but should honor the individual interrupt enables
```

That means that `wfi` resumes if individual interrupts are not masked in the `mie` register, and because `mstatus.MIE` does _not_ affect `wfi`, we can create an atomic section around it by _clearing_ `mstatus.MIE`. In the most basic form, we do a "global interrupt disable" on the hart, issue `wfi`, then do a "global interrupt enable" after coming out of `wfi`.

I'm not 100% sure what the privileged spec writers intended, but my best guess from reading is that the `mip` and `mie` CSRs are not supposed to be affected by `mstatus.MSIE`, and `mstatus.MSIE` only controls whether traps actually occur when something in `mip` is set.

Reply:
Ah, sorry, that was a little vague. For more of a template...

```c
// Disable interrupt handling while checking for pending work
irq_global_ctrl(false);
if (!work_is_pending) {
  // Will return whenever interrupts are pending in mip
  wait_for_interrupt();
}
// Enable interrupt handling to immediately service them
irq_global_ctrl(true);
// Now do any deferred work
process_pending_work();
```

Reply:
haha yeah.. i was originally going to use irq_global_ctrl...but felt that was a bit overkill. 
But that was the only way I could see getting around the corner case of interrupt coming in right between instructions.

Reply:
i ended up just throwing in the `irq_global_ctrl`... seems to easiest way to address this.